plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3BlobContainer.java

```diff
@@ -241,37 +241,23 @@ public void readBlobAsync(String blobName, ActionListener<ReadContext> listener)
                     return;
                 }
 
-                final List<CompletableFuture<InputStreamContainer>> blobPartInputStreamFutures = new ArrayList<>();
+                final List<ReadContext.StreamPartCreator> blobPartInputStreamFutures = new ArrayList<>();
                 final long blobSize = blobMetadata.objectSize();
                 final Integer numberOfParts = blobMetadata.objectParts() == null ? null : blobMetadata.objectParts().totalPartsCount();
                 final String blobChecksum = blobMetadata.checksum().checksumCRC32();
 
                 if (numberOfParts == null) {
-                    blobPartInputStreamFutures.add(getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, null));
+                    blobPartInputStreamFutures.add(() -> getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, null));
                 } else {
                     // S3 multipart files use 1 to n indexing
                     for (int partNumber = 1; partNumber <= numberOfParts; partNumber++) {
-                        blobPartInputStreamFutures.add(getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, partNumber));
+                        final int innerPartNumber = partNumber;
+                        blobPartInputStreamFutures.add(
+                            () -> getBlobPartInputStreamContainer(s3AsyncClient, bucketName, blobKey, innerPartNumber)
+                        );
                     }
```

Comment:
can we please test with this with restoring more number of shards at once ? The potential problem i see with this approach is the high number of connections we are opening . If we open it across shards and they get rate limited/thread pool starved for some time, the connections pool would get exhausted.

Reply:
So when does connection get opened exactly ? I might be wrong here if we open connections lazily , 20 at a time max per file.

Reply:
It is 20 per file, but we restore one file at a time. So in effect it is 20 at a time per shard.

Reply:
> Curious what is the throughput 20 streams can drive?

In my test on a `r7a.xlarge` node I was able to get an effective throughput of about 333 MiB/s.

I'm adding a new dynamic recovery setting called `indices.recovery.max_concurrent_remote_store_streams` for this parameter. The existing `cluster.routing.allocation.node_concurrent_recoveries` and `indices.recovery.max_bytes_per_sec` settings are being honored here as well.